### PR TITLE
bpo-37221: Add PyCode_NewWithPosOnlyArgs to be used internally and set PyCode_New as a compatibility wrapper

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue37221

### CODE REVIEW

**Comment by vstinner** [comment]
I would prefer to call the new function PyCode_NewWithPosArgs() rather than PyCode_NewEx().

**Comment by vstinner** [comment]
Would it make sense to wait to merge this PR just before next beta2, and also synchronize with Cython to make sure that Cython will be compatibl with Python 3.8 beta2, soon after beta2 is released?

**Comment by pablogsal** [comment]
I renamed the new function to `PyCode_NewWithPosOnlyArgs`

**Comment by pablogsal** [comment]
> Would it make sense to wait to merge this PR just before next beta2, and also synchronize with Cython to make sure that Cython will be compatible with Python 3.8 beta2, soon after beta2 is released?

We can merge and backport it and it will be released before in beta2, right? I will do the PR for Cython right away. They can chose to merge it when they think is better.

File: Doc/c-api/code.rst [hunk 1]
```diff
@@ -33,20 +33,22 @@ bound into a function.
 
    Return the number of free variables in *co*.
 
-.. c:function:: PyCodeObject* PyCode_New(int argcount, int posonlyargcount, int kwonlyargcount, int nlocals, int stacksize, int flags, PyObject *code, PyObject *consts, PyObject *names, PyObject *varnames, PyObject *freevars, PyObject *cellvars, PyObject *filename, PyObject *name, int firstlineno, PyObject *lnotab)
+.. c:function:: PyCodeObject* PyCode_NewWithPosOnlyArgs(int argcount, int posonlyargcount, int kwonlyargcount, int nlocals, int stacksize, int flags, PyObject *code, PyObject *consts, PyObject *names, PyObject *varnames, PyObject *freevars, PyObject *cellvars, PyObject *filename, PyObject *name, int firstlineno, PyObject *lnotab)
```
Comment (vstinner) [inline Doc/c-api/code.rst]: I suggest to move it *after* PyCode_New().

File: Doc/c-api/code.rst [hunk 2]
```diff
@@ -33,20 +33,22 @@ bound into a function.
...
 
    Return a new code object.  If you need a dummy code object to
    create a frame, use :c:func:`PyCode_NewEmpty` instead.  Calling
    :c:func:`PyCode_New` directly can bind you to a precise Python
    version since the definition of the bytecode changes often.
```
Comment (vstinner) [inline Doc/c-api/code.rst]: I suggest to write a simpler doc, like:

Similar to :c:func:`PyCode_New`, but with an extra "posonlyargcount" for positonal-only arguments.

File: Doc/whatsnew/3.8.rst [hunk 3]
```diff
@@ -952,6 +952,12 @@ Build and C API Changes
   allocation or deallocation may need to be adjusted.
   (Contributed by Eddie Elizondo in :issue:`35810`.)
 
+* The new function :c:func:`PyCode_NewWithPosOnlyArgs` substitutes
+  :c:func:`PyCode_New` as the prefered way to create code objects, having an
+  extra parameter representing the number of positional only parameters
+  (**posonlyargcount**).  The old function is kept for backward compatibility
+  reasons and will set the number os positional only parameters to 0.
```
Comment (vstinner) [inline Doc/whatsnew/3.8.rst]: I'm not sure about making PyCode_NewWithPosOnlyArgs the preferred way. PyCode_New() is fine for most use cases, no?

File: Objects/codeobject.c [hunk 4]
```diff
@@ -102,14 +102,27 @@ intern_string_constants(PyObject *tuple)
     return modified;
 }
 
-
 PyCodeObject *
-PyCode_New(int argcount, int posonlyargcount, int kwonlyargcount,
+PyCode_New(int argcount, int kwonlyargcount,
```
Comment (vstinner) [inline Objects/codeobject.c]: nitpick: *here* i suggest to declare PyCode_New *after* PyCode_NewWithPosOnlyArgs declaration.

File: Doc/c-api/code.rst [hunk 5]
```diff
@@ -33,20 +33,24 @@ bound into a function.
...
-      including positional-only.
+   .. audit-event:: code.__new__ "code filename name argcount kwonlyargcount nlocals stacksize flags"
+
+.. c:function:: PyCodeObject* PyCode_NewWithPosOnlyArgs(int argcount, int posonlyargcount, int kwonlyargcount, int nlocals, int stacksize, int flags, PyObject *code, PyObject *consts, PyObject *names, PyObject *varnames, PyObject *freevars, PyObject *cellvars, PyObject *filename, PyObject *name, int firstlineno, PyObject *lnotab)
+
+   Similar to :c:func:`PyCode_New`, but with an extra "posonlyargcount" for positonal-only arguments.
+
+   .. versionadded:: 3.8
 
    .. audit-event:: code.__new__ "code filename name argcount posonlyargcount kwonlyargcount nlocals stacksize flags"
```
Comment (vstinner) [inline Doc/c-api/code.rst]: I suggest to remove the audit event here, it's already documented in PyCode_New().

File: Doc/c-api/code.rst [hunk 6]
```diff
@@ -33,20 +33,24 @@ bound into a function.
...
+   Return a new code object.  If you need a dummy code object to create a frame,
+   use :c:func:`PyCode_NewEmpty` instead.  Calling :c:func:`PyCode_New` directly
+   can bind you to a precise Python version since the definition of the bytecode
+   changes often.
 
-   .. versionchanged:: 3.8
-      An extra parameter is required (*posonlyargcount*) to support :PEP:`570`.
-      The first parameter (*argcount*) now represents the total number of positional arguments,
-      including positional-only.
+   .. audit-event:: code.__new__ "code filename name argcount kwonlyargcount nlocals stacksize flags"
```
Comment (vstinner) [inline Doc/c-api/code.rst]: That's wrong, it's really `   .. audit-event:: code.__new__ "code filename name argcount posonlyargcount kwonlyargcount nlocals stacksize flags"` in practice (with pos only).

File: Doc/whatsnew/3.8.rst [hunk 7]
```diff
@@ -952,6 +952,10 @@ Build and C API Changes
   allocation or deallocation may need to be adjusted.
   (Contributed by Eddie Elizondo in :issue:`35810`.)
 
+* The new function :c:func:`PyCode_NewWithPosOnlyArgs` allows to create
+  code objects like :c:func:`PyCode_New`, but with an extra "posonlyargcount"
```
Comment (vstinner) [inline Doc/whatsnew/3.8.rst]: nitpick: `*posonlyargcount*` formatting

File: Include/code.h [hunk 8]
```diff
@@ -119,7 +119,14 @@ PyAPI_DATA(PyTypeObject) PyCode_Type;
 #define PyCode_GetNumFree(op) (PyTuple_GET_SIZE((op)->co_freevars))
 
 /* Public interface */
+
+/* For backwards compatibility after adding posonlyargcount */
```
Comment (vstinner) [inline Include/code.h]: IMHO this comment can be removed. There is no **need** to deprecate anything so far.

File: Misc/NEWS.d/next/C API/2019-06-11-02-50-38.bpo-37221.4tClQT.rst [hunk 9]
```diff
@@ -0,0 +1,5 @@
+The new function :c:func:`PyCode_NewWithPosOnlyArgs` substitutes
+:c:func:`PyCode_New` as the prefered way to create code objects, having an
+extra parameter representing the number of positional only parameters
+(**posonlyargcount**).  The old function is kept for backward compatibility
+reasons and will set the number os positional only parameters to 0.
```
Comment (vstinner) [inline Misc/NEWS.d/next/C API/2019-06-11-02-50-38.bpo-37221.4tClQT.rst]: I'm not sure that it's needed to deprecate/prefer anything so far. Just mention that you add the function, no?

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Include/code.h [hunk 10]
```diff
@@ -119,7 +119,13 @@ PyAPI_DATA(PyTypeObject) PyCode_Type;
 #define PyCode_GetNumFree(op) (PyTuple_GET_SIZE((op)->co_freevars))
 
 /* Public interface */
+
```
Comment (vstinner) [inline Include/code.h]: unrelated change

File: Doc/whatsnew/3.8.rst [hunk 11]
```diff
@@ -952,6 +952,10 @@ Build and C API Changes
   allocation or deallocation may need to be adjusted.
   (Contributed by Eddie Elizondo in :issue:`35810`.)
 
+* The new function :c:func:`PyCode_NewWithPosOnlyArgs` allows to create
+  code objects like :c:func:`PyCode_New`, but with an extra *posonlyargcount*
+  argument for indicating the number of positonal-only arguments.
```
Comment (vstinner) [inline Doc/whatsnew/3.8.rst]: extra argument or extra parameter?

File: Misc/NEWS.d/next/C API/2019-06-11-02-50-38.bpo-37221.4tClQT.rst [hunk 12]
```diff
@@ -0,0 +1,5 @@
+The new function :c:func:`PyCode_NewWithPosOnlyArgs` allows to create
+code objects like :c:func:`PyCode_New`, but with an extra *posonlyargcount*
+argument for indicating the number of positonal-only arguments.
+
+
```
Comment (vstinner) [inline Misc/NEWS.d/next/C API/2019-06-11-02-50-38.bpo-37221.4tClQT.rst]: you might remove empty lines :)))

**Review by vstinner — APPROVED** [review]
LGTM, but I would prefer a second review. @methane, @serhiy-storchaka, @ncoghlan, @encukou, anyone?

**Comment by scoder** [comment]
> Would it make sense to wait to merge this PR just before next beta2

Yes, please. Otherwise, we will loose a month of CI time during which Cython cannot adapt (because the CPython API version does not change before the release) and during which users cannot properly test with the unreleased CPython master because Cython does not work on it any more.

**Review by serhiy-storchaka — COMMENTED** [review]
Please add a What's New entry.

**Comment by serhiy-storchaka** [comment]
Would be nice if Cython release a new version which will use a special case only for Python `== 3.8b1`, not for `>= 3.8b1`.

**Comment by pablogsal** [comment]
@serhiy-storchaka There is already a what's new entry. Do you mean something extra or something different?

**Comment by serhiy-storchaka** [comment]
Sorry, I missed it.

**Review by ncoghlan — APPROVED** [review]
Thanks @pablogsal

**Comment by ncoghlan** [comment]
@serhiy-storchaka The issue for Cython is that the 3.8 dev branch is currently still mostly advertising itself as 3.8b1: https://github.com/python/cpython/blob/3.8/Include/patchlevel.h

@ambv Could we do something a little odd here, and change the branch version string to be "3.8b2-" and bump the release serial to "2" before the actual release? That way the change could be merged, and Cython would be able to correctly detect it as being a different build from 3.8b1.

**Comment by scoder** [comment]
> change the branch version string

Note that the dependency is in the generated C code and based on PY_VERSION_HEX, which cannot easily be modified in this way.

**Comment by ncoghlan** [comment]
@scoder I've posted a PR at https://github.com/cython/cython/pull/3009 which adds a check directly for the new PyCode_NewWithPosOnlyArgs symbol, which should let Cython reliably detect versions where the incompatible API change has been reverted, even before the Python 3.8.0b2 version bump.

If that's acceptable on the Cython side, then the only blocker to merging this on the CPython side would be fixing the merge conflict in What's New.

**Review by scoder — APPROVED** [review]

**Comment by pablogsal** [comment]
I have rebased to fix the merge conflicts.

**Comment by ambv** [comment]
> @ambv Could we do something a little odd here, and change the branch version string to be "3.8b2-" and bump the release serial to "2" before the actual release?

Sorry, missed the notification on this one. I'd rather not, let's just merge this shortly before beta2, say: Friday.

**Comment by encukou** [comment]
I read that as an invitation for anyone to hit the green button on Friday. Let us know if you want to do the merge yourself.

**Comment by scoder** [comment]
FYI, Cython 0.29.11 has been released with the necessary change for beta-2.

**Comment by vstinner** [comment]
> FYI, Cython 0.29.11 has been released with the necessary change for beta-2.

That's great! Thank you.

**Pull request merged**